### ISSUE TITLE
Tensorflow 2.0 keras load_model does not restore step / epoch counters

### ISSUE THREAD

**Issue by andreas-eberle** [tensorflow/tensorflow#27861]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes, see example below
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0 alpha 0 - gpu
- Python version: 3.5.3
- CUDA/cuDNN version: Cuda 10
- GPU model and memory: GeForce 960M 

**Describe the current behavior**
I train a keras model and let it be saved by the `keras.callbacks.ModelCheckpoint`. After the training is finished, I want to continue it from the saved checkpoint. Therefore, I load the model with `keras.models.load_model` and run `fit` again. Unfortunately, it seems that the steps/epochs counters are not restored. 

**Describe the expected behavior**
The steps /epochs counters should be restored to allow continuation of training. If they are not restored, e.g. tensorboard cannot be used properly, since the new training will write its values to the same steps as the first training, instead of appending them. The result looks like this: 

![Unbenannt](https://user-images.githubusercontent.com/9267365/56139393-12b50900-5f99-11e9-9c4e-b553ab2c1589.PNG)

**Code to reproduce the issue**
Run the following code multiple times and have a look at the tensorboard results by running `tensorboard --logdir=test_outputs`.

```
import os
from tensorflow import keras

log_dir = 'test_outputs'
model_file = os.path.join(log_dir, "model.hdf5")

(x_train, y_train), _ = keras.datasets.mnist.load_data()


def create_model():
	model = keras.models.Sequential([
		keras.layers.Flatten(input_shape=(28, 28), name="flatten"),
		keras.layers.Dense(128, activation='relu', name="dense1"),
		keras.layers.Dropout(0.2, name="dropout"),
		keras.layers.Dense(10, activation='softmax', name="dense2")
	])

	model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
	return model


if os.path.exists(model_file):
	model = keras.models.load_model(model_file)
else:
	model = create_model()

model.fit(x_train, y_train, epochs=5, callbacks=[keras.callbacks.TensorBoard(log_dir=log_dir), keras.callbacks.ModelCheckpoint(model_file)])
```

**Comment by unsatcore** [comment]
FWIW
it's possible to work around with model.fit (initial_epoch = ...)
obviously initial_epoch has to be stored somewhere - custom callback or as a part of checkpoint filename

**Comment by karmel** [comment]
This is the intended behavior; step and epoch are reset with call to model.fit. Optimizer.iterations should maintain its count, such that any learning rate adjustments reflect the model's history, but epoch count is designed to restart with new training loop, even when loaded from a serialized model.

**Comment by andreas-eberle** [comment]
But how can I then continue training a model and see the correct graph in Tensorboard? If I restore a checkpoint, I expect to be able to continue working with it. This is a crucial feature!

@karmel: Can you please explain, how we can handle this need?

**Comment by jackd** [comment]
@karmel `Model.fit` allows an `initial_epoch` argument. Surely this is evidence that the epoch count is designed to support restarting from an existing checkpoint, even if it defaults to restarting from 0?

@andreas-eberle I've had similar frustrations between this and `ModelCheckpoint`. I've ended up writing my own saving/restoration callback and (for the moment) hacking the `TensorBoard` callback to do what is, in my opinion, "the right thing", but an independent implementation is probably wise at some point to ensure forwards compatibility (I'm accessing private attrs).

It's hacky and gross and written for eager execution. There are slight issues here surrounding the final element of each epoch which might have fewer members than `batch_size`, but you might be able to adapt it enough to be suitable for your purposes.

```python
class CheckpointManagerCallback(tf.keras.callbacks.Callback):
    """
    Callback wraping `tf.train.CheckpointManager`.

    Restores previous checkpoint `on_train_begin`

    Example usage:
    ```python
    model = get_model(...)
    model.compile(optimizer=optimizer, ...)
    checkpoint = tf.train.Checkpoint(optimizer=optimizer, model=model)
    manager = tf.train.CheckpointManager(
        checkpoint, '/tmp/my_model', max_to_keep=5)
    callback = CheckpointManagerCallback(checkpoint, manager, period=1)

    model.fit(..., callbacks=[callbacks])
    ```
    """
    def __init__(self, checkpoint, manager, period=1, save_on_train_end=True):
        self._manager = manager
        self._checkpoint = checkpoint
        self._period = period
        self._save_on_train_end = save_on_train_end
        self._restored = False
        self._epoch_count = None
        self._last_save = None

    def _on_begin(self):
        if not self._restored:
            self.restore()

    def restore(self, save_path=None):
        if save_path is None:
            save_path = self._manager.latest_checkpoint
        self._checkpoint.restore(save_path)
        self._restored = True

    def on_train_begin(self, logs=None):
        self._on_begin()

    def on_test_begin(self, logs=None):
        self._on_begin()

    def on_predict_begin(self, logs=None):
        self._on_begin()

    def on_epoch_end(self, epoch, logs=None):
        epochs_finished = epoch + 1
        self._epoch_count = epochs_finished
        if epochs_finished % self._period == 0:
            self._save()

    def on_train_end(self, logs=None):
        if self._save_on_train_end:
            self._save()

    def _save(self):
        if self._epoch_count is None:
            return
        if self._last_save != self._epoch_count:
            self._manager.save(self._epoch_count)
            self._last_save = self._epoch_count

checkpoint = tf.train.Checkpoint(
    model=model, optimizer=model.optimizer)
manager = tf.train.CheckpointManager(
    checkpoint, model_dir, max_to_keep=max_ckpts_to_keep)
saver_callback = cb.CheckpointManagerCallback(
    checkpoint, manager, period=checkpoint_freq)

saver_callback.restore()

tb_callback = tf.keras.callbacks.TensorBoard(
    write_graph=False, log_dir=model_dir, update_freq=summary_freq)

tb_callback._total_batches_seen = initial_train_steps
tb_callback._samples_seen = initial_train_steps * batch_size
if val_steps_per_epoch is not None:
    initial_val_steps = initial_epoch*val_steps_per_epoch
    tb_callback._total_val_batches_seen = initial_val_steps

model.fit(callbacks=[saver_callback, tb_callback], initial_epoch=initial_epoch, ...)
```

Coming from the `estimator` API it sure would be nice if keras had something close to feature parity at least by 2.0. Obvious issues like this being labelled as "working as intended" doesn't give me much confidence...

**Comment by fzyzcjy** [comment]
Also think that the behavior is not intended... For instance, many *learning rate schedulers* will look at the *epoch* number to adjust the learning rate. So when continuing from a checkpoint, the learning rate will be completely wrong...

**Comment by yourtheron** [comment]
I have the same issue. Restoring current epoch is important on some occasions. For example, if I use L1 cost in the first 1000 epochs, and turn into L2 cost afterwards. When I restore a ckpt, I'd like to know how many epochs have elapsed.

**Comment by hoenickf** [comment]
Please reopen. It would be great to at least define the current epoch and step count. Otherwise the `tf.keras.callbacks.ModelCheckpoint(f'{self.checkpoint_dir}/{{epoch}}')`
overwrites the existing checkpoints.

**Comment by JustASquid** [comment]
+1, this seems like a bizarrely missing feature that renders training resumption impractical for real use cases, at least without cumbersome custom code/callbacks to manually save and restore the epoch.
